agent/src/beerocks/slave/link_metrics/ieee802_3_link_metrics_collector.cpp

```diff
@@ -0,0 +1,246 @@
+                                                                        speed)) {
+                    if (SPEED_100 == speed) {
+                        phy_rate = 100;
+                    } else if (SPEED_1000 == speed) {
+                        phy_rate = 1000;
+                    }
+                }
+
+                link_metrics.transmitter.packet_errors       = stats->tx_errors;
+                link_metrics.transmitter.transmitted_packets = stats->tx_packets;
+                /**
+                 * Note: The MAC throughput capacity is a function of the physical data rate and
+                 * of the MAC overhead. We could somehow compute such overhead or, for simplicity,
+                 * set the MAC throughput as a percentage of the physical data rate. To make
+                 * things even simpler, we will set the MAC throughput capacity with the same
+                 * value as the physical data rate, as if there was no overhead at all.
+                 */
+                link_metrics.transmitter.mac_throughput_capacity = phy_rate;
+                // Note: For simplicity, link availability is set to "100% of the time"
+                link_metrics.transmitter.link_availability = 100;
+                link_metrics.transmitter.phy_rate          = phy_rate;
+
+                link_metrics.receiver.packet_errors    = stats->rx_errors;
+                link_metrics.receiver.packets_received = stats->rx_packets;
+                link_metrics.receiver.rssi             = UINT8_MAX;
+            }
+            break;
+        }
+    }
+
+    return result;
+}
+
+static bool get_link_metrics(int fd, const std::string &local_interface_name,
+                             sLinkMetrics &link_metrics)
+{
+    bool result = false;
+
+    struct sockaddr_nl kernel; /* the remote (kernel space) side of the communication */
+
+    struct msghdr rtnl_msg; /* generic msghdr struct for use with sendmsg */
+    struct iovec io;        /* IO vector for sendmsg */
+    struct nl_req_t req;    /* structure that describes the Netlink packet itself */
+
+    /**
+     * Netlink socket is ready for use, prepare and send request
+     */
+    memset(&rtnl_msg, 0, sizeof(rtnl_msg));
+    memset(&kernel, 0, sizeof(kernel));
+    memset(&req, 0, sizeof(req));
+
+    kernel.nl_family = AF_NETLINK; /* fill-in kernel address (destination of our message) */
+
+    req.hdr.nlmsg_len    = NLMSG_LENGTH(sizeof(struct rtgenmsg));
+    req.hdr.nlmsg_type   = RTM_GETLINK;
+    req.hdr.nlmsg_flags  = NLM_F_REQUEST | NLM_F_DUMP;
+    req.hdr.nlmsg_seq    = 1;
+    req.hdr.nlmsg_pid    = 0;
+    req.gen.rtgen_family = AF_PACKET; /*  no preferred AF, we will get *all* interfaces */
+
+    io.iov_base          = &req;
+    io.iov_len           = req.hdr.nlmsg_len;
+    rtnl_msg.msg_iov     = &io;
+    rtnl_msg.msg_iovlen  = 1;
+    rtnl_msg.msg_name    = &kernel;
+    rtnl_msg.msg_namelen = sizeof(kernel);
```

Comment:
I agree with you. 

This is something that could probably be done using changes in "Feature/add support for nl80211 standard commands to BWL #900". However, be aware that such changes have been made into BWL and not in network_utils. 

Anyway, I agree with you that this is a refactoring that could be done later (I think it is not going to be so easy and quick as might seem at first glance).

Reply:
Why not indeed move `base_wlan_hal_nl80211::send_nl80211_msg` to BCL? just that function, shouldn't be too hard even as a followup commit in this PR. I really prefer that we don't duplicate code when we really don't need to. Still, as you say, it can be done later, so (nitpoick)

Reply:
Method base_wlan_hal_nl80211::send_nl80211_msg() is going to disappear after we had approved the design and merged "Feature/add support for nl80211 standard commands to BWL #900" and done a refactoring of the NL80211 version of the BWL. 

The work to be done is then to implement the get_link_metrics() method using the new classes defined in #900.